### ISSUE TITLE
Support sparse creates.io registry

### ISSUE THREAD

**Issue by Milo123459** [cargo-public-api/cargo-public-api#408]
When enabling sparse registry, I consistently got this error:
```
Error: The repo at path /home/codespace/.cargo/registry/index/github.com-1ecc6299db9ec823 is unusable due to having an invalid HEAD reference: reference 'refs/heads/main' not found; class=Reference (4); code=NotFound (-3)
```
Once disabled, and recreating the registry (rming it then cargo fetch) it works.

Command: `cargo public-api diff latest`

**Comment by Enselic** [comment]
Thanks a lot for reporting. I can reproduce with our git repo like this. Note how I get a different error first:
```
$ export CARGO_REGISTRIES_CRATES_IO_PROTOCOL=sparse

$ rm -rf ~/.cargo/registry

$ cargo run -- -p rustup-toolchain diff latest
[...]
Error: there is no TLS stream available; class=Ssl (16)
Caused by:
    there is no TLS stream available; class=Ssl (16)

$ cargo run -- -p rustup-toolchain diff latest
[...]
Error: The repo at path /home/martin/.cargo/registry/index/github.com-1ecc6299db9ec823 is unusable due to having an invalid HEAD reference: reference 'refs/heads/master' not found; class=Reference (4); code=NotFound (-3)
```

Looks like we either need to wait for https://github.com/frewsxcv/rust-crates-index/issues/87 or roll our own solution. Personally I would wait, but if someone comes up with a nice and simple way of doing it ourselves, I would be open to merge that until our deps support it.  But honestly it would be much better to put that effort into adding support upstream.

**Comment by Milo123459** [comment]
I also saw that error first, then after a bit, it went away and switched to the one I sent in the issue.

**Comment by Enselic** [comment]
The registry is only used when diffing. More specifically, it is used to create a "fake" Cargo project so that the rustdoc JSON for an old version can be built. I wonder if maybe we should always use the "sparse" index to get what we need for creating the "fake" Cargo.toml, and then let cargo use whatever it wants. So maybe we can do something like this:

```console
$ curl https://index.crates.io/ex/am/example_api
{"name":"example_api","vers":"0.1.0","deps":[],"cksum":"d0f99c425515721620183a74993ebb809ca7f06c4e6b09b8c28668d56a2e78fd","features":{},"yanked":false}
{"name":"example_api","vers":"0.1.1","deps":[],"cksum":"3fc31d0e953c46287ea7ddb2f5cd93f0b8eae0ec9777f75569b9d3b7837c14bc","features":{},"yanked":false}
{"name":"example_api","vers":"0.2.0","deps":[],"cksum":"67c4a4db10e5e11217344e754b634eab454ee983a3123c576a2781baefc08bc8","features":{},"yanked":false}
{"name":"example_api","vers":"0.3.0","deps":[],"cksum":"8933479dd187e1d4d7de03e4dfd67e23d3ca494da830d1718710c779b5bd2156","features":{},"yanked":false}
{"name":"example_api","vers":"0.2.1","deps":[],"cksum":"2cb9ebe24cd5cdf18d63dd22cbc85fc6ad94aa10645038e8cfebead34186aaff","features":{"feature":[]},"yanked":false}
```

to get information about versions and features. Then we create a `Cargo.toml` (see https://github.com/Enselic/cargo-public-api/blob/main/cargo-public-api/src/published_crate.rs), and then cargo would use a sparse index if configured, or the git repo if configured.

With this approach we should be able to get rid of our [crates-index](https://crates.io/crates/crates-index) dependency, which pulls in like 30 additional deps. We just need some simple serde_json code to parse the above HTTP response. (But we can keep using `crates-index` as a first step.)

I'm a bit short on time, so I probably won't have time to try it out until earliest next week. But I think this could work quite well.

The downside would be that we would require an internet connection. But that is typically needed anyway when downloading old versions of crates. So I don't think we need to worry about that.

I'm tentatively removing the limitation-upstream label. Also setting high-prio since a sparse index is the default in Rust 1.70 and onwards.

**Comment by Enselic** [comment]
For visibility, I'll keep this open until I have made a release with the fix. I think that will happen within a week.

**Comment by Enselic** [comment]
Fixed in [0.31.2](https://crates.io/crates/cargo-public-api/0.31.2)
